src/platform/backends/shared/base_virtual_machine.cpp

```diff
@@ -17,20 +17,78 @@
 
 #include "base_virtual_machine.h"
 
+#include <multipass/exceptions/file_not_found_exception.h>
+#include <multipass/exceptions/snapshot_exceptions.h>
 #include <multipass/exceptions/ssh_exception.h>
+#include <multipass/file_ops.h>
 #include <multipass/logging/log.h>
+#include <multipass/snapshot.h>
+#include <multipass/top_catch_all.h>
+#include <multipass/utils.h>
+#include <multipass/vm_specs.h>
+
+#include <scope_guard.hpp>
+
+#include <QDir>
+
+#include <regex>
 
 namespace mp = multipass;
 namespace mpl = multipass::logging;
+namespace mpu = multipass::utils;
+
+namespace
+{
+using St = mp::VirtualMachine::State;
+constexpr auto snapshot_extension = "snapshot.json";
+constexpr auto head_filename = "snapshot-head";
+constexpr auto count_filename = "snapshot-count";
+constexpr auto yes_overwrite = true;
+
+void assert_vm_stopped(St state)
+{
+    assert(state == St::off || state == St::stopped);
+}
+
+mp::Path derive_head_path(const QDir& snapshot_dir)
+{
+    return snapshot_dir.filePath(head_filename);
+}
+
+void update_parents_rollback_helper(const std::shared_ptr<mp::Snapshot>& deleted_parent,
+                                    std::vector<mp::Snapshot*>& updated_parents)
+{
+    for (auto snapshot : updated_parents)
+        snapshot->set_parent(deleted_parent);
+}
+
+std::string trim(const std::string& s)
+{
+    return std::regex_replace(s, std::regex{R"(^\s+|\s+$)"}, "");
+}
```

Comment:
Yup. Both improvements coming.

Reply:
For this one I ended up falling into a bit of a rabbit hole. I saw that there was already a `trim_end` that was able to trim in place. I thought I'd better be consistent, reuse it, and avoid copying strings (which is nice on full file contents...). So I dropped the regex approach entirely. From there, it bugged me that I could not use the new utils with rvalues and went on to generalize those trimmers with templates, but I wanted to be careful to avoid any dangling refs or other pitfalls (I wouldn't mind another pair of attentive eyes on that code, just to be sure). The templates also allowed me to easily drop the `std::function` wrappers. Please let me know if you're good with it.

Reply:
First, one little mistake here
```cpp
template <typename Str, typename Filter>
Str&& multipass::utils::trim(Str&& s, Filter&& filter)
{
    return trim_begin(trim_end(std::forward<Str>(s), std::forward<Filter>(filter)));
}
```
I think `filter` should be passed to both `trim_begin` and `trim_end`. Now it is passed only to `trim_end` and `trim_begin` will always use `is_space`.

Also, there is one minuscule detail about
```cpp
inline constexpr auto is_space = static_cast<int (*)(int)>(std::isspace);
```
since [the docs](https://en.cppreference.com/w/cpp/string/byte/isspace#Notes) say that it's 'wrong' to use `static_cast<int (*)(int)>(std::isspace)` when using it on `std::string` and it should be instead used as `[](unsigned char c){ return std::isspace(c); }`. I'm not really sure if this matters much, but I'd like to hear your opinion as well.

Reply:
Ahh, I knew I needed you to have a look. Fixes coming.